### PR TITLE
feat: prefer newer file when `fetch`

### PR DESCRIPTION
**Pull request by SettingDust**

### CODE REVIEW

File: src/commonMain/kotlin/teksturepako/pakku/api/actions/fetch/Fetch.kt [hunk 1]
```diff
@@ -25,9 +25,7 @@ fun retrieveProjectFiles(
     lockFile: LockFile,
     providers: List<Provider>
 ) : List<Result<ProjectFile, ActionError>> = lockFile.getAllProjects().map { project ->
-    val file = providers.firstNotNullOfOrNull { provider ->
-        project.getFilesForProvider(provider).firstOrNull()
-    }
+    val file = project.getFilesForProviders(providers).maxByOrNull { it.datePublished }
```
Comment (juraj-hrivnak) [inline src/commonMain/kotlin/teksturepako/pakku/api/actions/fetch/Fetch.kt:28]: You could use the spread operator.

**Comment by juraj-hrivnak** [comment]
Also, you probably want to do it [here](https://github.com/juraj-hrivnak/Pakku/blob/main/src/commonMain/kotlin/teksturepako/pakku/api/actions/export/RuleContext.kt#L115-L117) and [here](https://github.com/juraj-hrivnak/Pakku/blob/main/src/commonMain/kotlin/teksturepako/pakku/api/actions/export/RuleContext.kt#L251-L253).

File: src/commonMain/kotlin/teksturepako/pakku/api/actions/fetch/Fetch.kt (hunk 1 above)
Comment (SettingDust) [reply src/commonMain/kotlin/teksturepako/pakku/api/actions/fetch/Fetch.kt:28]: convert to typed array?

File: src/commonMain/kotlin/teksturepako/pakku/api/projects/Project.kt [hunk 2]
```diff
@@ -182,6 +182,13 @@ data class Project(
         return this.files.filter { it.type in providers.map { provider -> provider.serialName } }
     }
 
+    fun getFilesForProviders(providers: Collection<Provider>): List<ProjectFile>
+    {
+        return this.files.filter { it.type in providers.map { provider -> provider.serialName } }
+    }
+
+    fun getLatestFile(providers: Collection<Provider>) = getFilesForProviders(providers).maxByOrNull { it.datePublished }
```
Comment (juraj-hrivnak) [inline src/commonMain/kotlin/teksturepako/pakku/api/projects/Project.kt:190]: Only a question, what would happen when the datePublished is the default value?

**Review by juraj-hrivnak — APPROVED** [review]

File: src/commonMain/kotlin/teksturepako/pakku/api/projects/Project.kt (hunk 2 above)
Comment (SettingDust) [reply src/commonMain/kotlin/teksturepako/pakku/api/projects/Project.kt:190]: natural order

**Pull request merged**